tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+                    {
+                        propagatedTagsBuilder.Length--;
+                    }
+
+                    propagatedTags = StringBuilderCache.GetStringAndRelease(propagatedTagsBuilder);
+                }
+                else
+                {
+                    propagatedTags = null;
+                }
+
+                traceState = new W3CTraceState(samplingPriority, origin, propagatedTags);
                 return true;
             }
+            finally
+            {
+                StringBuilderCache.Release(propagatedTagsBuilder);
+            }
+        }
+
+        [return: NotNullIfNotNull("samplingPriority")]
+        private static string? SamplingPriorityToString(int? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       2 => "2",
+                       1 => "1",
+                       0 => "0",
+                       -1 => "-1",
+                       null => null,
+                       not null => samplingPriority.Value.ToString(CultureInfo.InvariantCulture)
+                   };
+        }
 
-            return false;
+        private static int? SamplingPriorityToInt32(string? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       "2" => 2,
+                       "1" => 1,
+                       "0" => 0,
+                       "-1" => 1,
+                       null => null,
+                       not null => int.TryParse(samplingPriority, out var result) ? result : null
+                   };
         }
 
-        private static bool IsValidTraceId([NotNullWhen(true)] string? traceId)
+#if NETCOREAPP
+        private static int? SamplingPriorityToInt32(ReadOnlySpan<char> samplingPriority)
         {
-            if (string.IsNullOrEmpty(traceId))
+            if (samplingPriority.Equals("2", StringComparison.Ordinal))
             {
-                return false;
+                return 2;
             }
 
-            if (traceId!.Length != 32)
+            if (samplingPriority.Equals("1", StringComparison.Ordinal))
             {
+                return 1;
+            }
+
+            if (samplingPriority.Equals("0", StringComparison.Ordinal))
+            {
+                return 0;
+            }
+
+            if (samplingPriority.Equals("-1", StringComparison.Ordinal))
+            {
+                return -1;
+            }
+
+            return int.TryParse(samplingPriority, out var result) ? result : null;
+        }
+#endif
+
+        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
+            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        {
+            spanContext = null;
+
+            // get the "traceparent" header
+            var traceParentHeader = ParseUtility.ParseString(carrier, carrierGetter, TraceParentHeaderName)?.Trim();
```

Comment:
How are we supposed to handle multiple `traceparent` headers? This takes the first one it finds, is that correct?

Reply:
I need to update this. The W3C RFC specifies that multiple `traceparent` headers is invalid and we should ignore them all. We just updated our internal RFC to reflect this.

Reply:
This is now done. If there are multiple `traceparent` headers, we ignore them as invalid. If there are multuple `tracestate` headers, we join them as one (although we're still only reading the `dd=` value and ignoring everything else, separate issue).